File: src/ynab.ts
import * as ynab from "ynab";
import { TransactionClearedStatus } from "ynab";
import { z } from "zod";
import { Account, BankName, bankNames } from "./banks/types.js";
import env from "./utils/env.js";
import logger from "./utils/logger.js";
import secrets from "./utils/secrets.js";

const ynabAPI = new ynab.API(secrets.YNAB_ACCESS_TOKEN);

export async function importTransactions(accounts: z.infer<typeof Account>[]) {
  logger.debug("Importing transactions to YNAB");

  logger.debug("Fetching YNAB accounts");
  const ynabAccounts = await ynabAPI.accounts
    .getAccounts(env.YNAB_BUDGET_ID)
    .then((response) =>
      response.data.accounts.filter((account) => !account.deleted),
    );

  if (
    !accounts.some(
      (account) => account.transactions && account.transactions.length,
    )
  ) {
    logger.debug("Imported 0 transaction(s)");
    return;
  }

  const importedTransactionMap: Record<string, number> = {};

  const findYnabAccount = (account: z.infer<typeof Account>) => {
    return ynabAccounts.find((ynabAccount) =>
      ynabAccount.note?.includes(account.id),
    );
  };

  const transactionsToImport = accounts
    .filter((account) => account.transactions && account.transactions.length)
    .flatMap((account) => {
      const matchedYnabAccount = findYnabAccount(account);

      if (!matchedYnabAccount) {
        return [];
      }

      return account.transactions
        .filter((transaction) => {
          const transactionDate = new Date(transaction.date);
          const now = new Date();
          const fiveYearsAgo = new Date();
          fiveYearsAgo.setFullYear(now.getFullYear() - 5);
          return transactionDate <= now && transactionDate >= fiveYearsAgo;
        })
        .map((transaction) => {
          const amount = Math.round(transaction.amount * 1000);

          const key = `${matchedYnabAccount.id}:${amount}:${transaction.date}`;
          if (importedTransactionMap[key]) {
            importedTransactionMap[key]++;
          } else {
            importedTransactionMap[key] = 1;
          }

          return {
            account_id: matchedYnabAccount.id,
            date: transaction.date,
            amount,
            payee_name: transaction.description,
            cleared: TransactionClearedStatus.Cleared,
            import_id: `YNAB:${amount}:${transaction.date}:${importedTransactionMap[key]}`,
          };
        });
    });

  if (!transactionsToImport.length) {
    logger.debug("Imported 0 transaction(s)");
    return;
  }

  const response = await ynabAPI.transactions.createTransactions(
    env.YNAB_BUDGET_ID,
    { transactions: transactionsToImport },
  );

  const { transactions: imported } = response.data;
  logger.debug(`Imported ${imported?.length} transaction(s)`, imported);
}

export async function updateAccountBalances(
  accounts: z.infer<typeof Account>[],
  bankName: BankName,
) {
  logger.debug("Updating account balances in YNAB");

  logger.debug("Fetching YNAB accounts");
  const ynabAccounts = await ynabAPI.accounts
    .getAccounts(env.YNAB_BUDGET_ID)
    .then((response) =>
      response.data.accounts.filter((account) => !account.deleted),
    );

  logger.debug(`Fetched ${ynabAccounts.length} YNAB accounts`);

  let adjustmentsCreated = 0;

  const findYnabAccount = (account: z.infer<typeof Account>) => {
    return ynabAccounts.find((ynabAccount) =>
      ynabAccount.note?.includes(account.id),
    );
  };

  for (const account of accounts) {
    const matchedYnabAccount = findYnabAccount(account);

    if (!matchedYnabAccount) {
      logger.debug(
        `No matching YNAB account found for bank account ${account.id}`,
      );
      continue;
    }

    const desiredBalance = Math.round(account.balance * 1000);
    const currentBalance = matchedYnabAccount.balance;
    const adjustment = desiredBalance - currentBalance;

    if (adjustment !== 0) {
      const transaction = {
        account_id: matchedYnabAccount.id,
        payee_id: "df22f16d-449b-4214-98da-2844de326faf",
        date: new Date().toLocaleDateString("en-CA", {
          timeZone: "America/Halifax",
        }),
        amount: adjustment,
        memo: `Entered automatically from ${bankNames[bankName]}`,
        cleared: TransactionClearedStatus.Reconciled,
        approved: true,
      };

      await ynabAPI.transactions.createTransaction(env.YNAB_BUDGET_ID, {
        transaction,
      });

      adjustmentsCreated++;
      logger.debug(
        `Created balance adjustment for YNAB account ${matchedYnabAccount.id} (${matchedYnabAccount.name}): ${adjustment} milliunits`,
      );
    } else {
      logger.debug(
        `No adjustment needed for YNAB account ${matchedYnabAccount.id} (${matchedYnabAccount.name})`,
      );
    }
  }

  logger.debug(`Created ${adjustmentsCreated} balance adjustment(s)`);
}

File: src/ynab.py
import ynab
from ynab import TransactionClearedStatus
from pydantic import BaseModel
from typing import List
from .banks.types import Account
from .utils.env import env
from .utils.logger import logger
from .utils.secrets import secrets

ynabAPI = ynab.API(secrets.YNAB_ACCESS_TOKEN)

class Transaction(BaseModel):
    account_id: str
    date: str
    amount: int
    payee_name: str
    cleared: str
    import_id: str

async def import_transactions(accounts: List[Account]):
    logger.debug("Importing transactions to YNAB")

    logger.debug("Fetching YNAB accounts")
    ynab_accounts = await ynabAPI.accounts.get_accounts(env.YNAB_BUDGET_ID)
    ynab_accounts = [account for account in ynab_accounts.data.accounts if not account.deleted]

    if not any(account.transactions for account in accounts):
        logger.debug("Imported 0 transaction(s)")
        return

    imported_transaction_map = {}

    def find_ynab_account(account: Account):
        return next((ynab_account for ynab_account in ynab_accounts if account.id in (ynab_account.note or "")), None)

    transactions_to_import = []
    for account in accounts:
        if not account.transactions:
            continue

        matched_ynab_account = find_ynab_account(account)
        if not matched_ynab_account:
            continue

        for transaction in account.transactions:
            transaction_date = datetime.strptime(transaction.date, "%Y-%m-%d")
            now = datetime.now()
            five_years_ago = now - timedelta(days=5*365)
            if not (five_years_ago <= transaction_date <= now):
                continue

            amount = round(transaction.amount * 1000)
            key = f"{matched_ynab_account.id}:{amount}:{transaction.date}"
            imported_transaction_map[key] = imported_transaction_map.get(key, 0) + 1

            transactions_to_import.append(Transaction(
                account_id=matched_ynab_account.id,
                date=transaction.date,
                amount=amount,
                payee_name=transaction.description,
                cleared=TransactionClearedStatus.Cleared,
                import_id=f"YNAB:{amount}:{transaction.date}:{imported_transaction_map[key]}"
            ))

    if not transactions_to_import:
        logger.debug("Imported 0 transaction(s)")
        return

    response = await ynabAPI.transactions.create_transactions(env.YNAB_BUDGET_ID, {"transactions": transactions_to_import})
    imported = response.data.transactions
    logger.debug(f"Imported {len(imported)} transaction(s)", imported)
